# Venture Funding with Deep Learning

You work as a risk management associate at Alphabet Soup, a venture capital firm. Alphabet Soup’s business team receives many funding applications from startups every day. This team has asked you to help them create a model that predicts whether applicants will be successful if funded by Alphabet Soup.

The business team has given you a CSV containing more than 34,000 organizations that have received funding from Alphabet Soup over the years. With your knowledge of machine learning and neural networks, you decide to use the features in the provided dataset to create a binary classifier model that will predict whether an applicant will become a successful business. The CSV file contains a variety of information about these businesses, including whether or not they ultimately became successful.

* Prepare the data for use on a neural network model.

* Compile and evaluate a binary classification model using a neural network.

* Optimize the neural network model.

### Prepare the Data for Use on a Neural Network Model 

1. Read the `applicants_data.csv` file into a Pandas DataFrame. Review the DataFrame, looking for categorical variables that will need to be encoded, as well as columns that could eventually define features and target variables.   

2. Drop the “EIN” (Employer Identification Number) and “NAME” columns from the DataFrame, because they are not relevant to the binary classification model.
 
3. Encode the dataset’s categorical variables using `OneHotEncoder`, and then place the encoded variables into a new DataFrame.

4. Add the original DataFrame’s numerical variables to the DataFrame containing the encoded variables.

5. Using the preprocessed data, create the features (`X`) and target (`y`) datasets. The target dataset should be defined by the preprocessed DataFrame column “IS_SUCCESSFUL”. The remaining columns should define the features dataset. 

6. Split the features and target sets into training and testing datasets.

7. Use scikit-learn's `StandardScaler` to scale the features data.

### Compile and Evaluate a Binary Classification Model Using a Neural Network

Use TensorFlow to design a binary classification deep neural network model. This model should use the dataset’s features to predict whether an Alphabet Soup&ndash;funded startup will be successful based on the features in the dataset.  Then, compile and fit your model. Finally, evaluate the binary classification model to calculate the model’s loss and accuracy. 
 
To do so, complete the following steps:

1. Create a deep neural network by assigning the number of input features, the number of layers, and the number of neurons on each layer using Tensorflow’s Keras.

2. Compile and fit the model using the `binary_crossentropy` loss function, the `adam` optimizer, and the `accuracy` evaluation metric.

3. Evaluate the model using the test data to determine the model’s loss and accuracy.

### Optimize the Neural Network Model

Using TensorFlow and Keras, optimize the model to improve the model's accuracy. 

1. Define at least three new deep neural network models (the original plus 2 optimization attempts). With each, try to improve on your first model’s predictive accuracy.

> **Rewind** Recall that perfect accuracy has a value of 1, so accuracy improves as its value moves closer to 1. To optimize your model for a predictive accuracy as close to 1 as possible, you can use any or all of the following techniques:
>
> * Adjust the input data by dropping different features columns to ensure that no variables or outliers confuse the model.
>
> * Add more neurons (nodes) to a hidden layer.
>
> * Add more hidden layers.
>
> * Use different activation functions for the hidden layers.
>
> * Add to or reduce the number of epochs in the training regimen.

2. Display the accuracy scores achieved by each model, and compare the results.
---

In [2]:
# Imports
import pandas as pd
from pathlib import Path
import tensorflow as tf
from tensorflow.keras.layers import Dense
from tensorflow.keras.models import Sequential
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler,OneHotEncoder

---

## Step 1: Prepare the data to be used on a neural network model

In [3]:
# Read the applicants_data.csv file into a file path, verified the path exists
# (in markdown), and then created a Pandas DataFrame from the csv file
file_path = Path("../fintech-boot-camp/Homework/13-Neural-Networks/Instructions/Starter_Code/Resources/applicants_data.csv")
#file_path.exists()
#print("file exists!")
      
applicant_data_df = pd.read_csv(file_path)

In [4]:
# Review the dataframe, and the data types associated with the columns
applicant_data_df
#applicant_data_df.dtypes

EIN                                               NAME  \
0       10520599                       BLUE KNIGHTS MOTORCYCLE CLUB   
1       10531628             AMERICAN CHESAPEAKE CLUB CHARITABLE TR   
2       10547893                 ST CLOUD PROFESSIONAL FIREFIGHTERS   
3       10553066                     SOUTHSIDE ATHLETIC ASSOCIATION   
4       10556103           GENETIC RESEARCH INSTITUTE OF THE DESERT   
...          ...                                                ...   
34294  996009318              THE LIONS CLUB OF HONOLULU KAMEHAMEHA   
34295  996010315           INTERNATIONAL ASSOCIATION OF LIONS CLUBS   
34296  996012607                                PTA HAWAII CONGRESS   
34297  996015768  AMERICAN FEDERATION OF GOVERNMENT EMPLOYEES LO...   
34298  996086871                           WATERHOUSE CHARITABLE TR   

      APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  \
0       Association       1              0                      N      5000   
1      Co-operative       1         1-9999                      N    108590   
2       Association       1              0                      N      5000   
3             Trust       1    10000-24999                      N      6692   
4             Trust       1  100000-499999                      N    142590   
...             ...     ...            ...                    ...       ...   
34294   Association       1              0                      N      5000   
34295   Association       1              0                      N      5000   
34296   Association       1              0                      N      5000   
34297   Association       1              0                      N      5000   
34298  Co-operative       1          1M-5M                      N  36500179   

       IS_SUCCESSFUL  
0                  1  
1                  1  
2                  0  
3                  1  
4                  1  
...              ...  
34294              0  
34295              0  
34296              0  
34297              1  
34298              0  

[34299 rows x 12 columns]

### Step 2: Drop the “EIN” (Employer Identification Number) and “NAME” columns from the DataFrame, because they are not relevant to the binary classification model.

In [5]:
# Drop the 'EIN' and 'NAME' columns from the DataFrame, they are irrelevant to
# the binary classification model
applicant_data_df = applicant_data_df.drop(columns=["EIN", "NAME"])

# Review the DataFrame
applicant_data_df

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  \
0       Association       1              0                      N      5000   
1      Co-operative       1         1-9999                      N    108590   
2       Association       1              0                      N      5000   
3             Trust       1    10000-24999                      N      6692   
4             Trust       1  100000-499999                      N    142590   
...             ...     ...            ...                    ...       ...   
34294   Association       1              0                      N      5000   
34295   Association       1              0                      N      5000   
34296   Association       1              0                      N      5000   
34297   Association       1              0                      N      5000   
34298  Co-operative       1          1M-5M                      N  36500179   

       IS_SUCCESSFUL  
0                  1  
1                  1  
2                  0  
3                  1  
4                  1  
...              ...  
34294              0  
34295              0  
34296              0  
34297              1  
34298              0  

[34299 rows x 10 columns]

### Step 3: Encode the dataset’s categorical variables using `OneHotEncoder`, and then place the encoded variables into a new DataFrame.

In [6]:
# Create a list of categorical variables 
categorical_variables = list(applicant_data_df.dtypes[
                             applicant_data_df.dtypes == "object"].index)

# Display the categorical variables list
categorical_variables

['APPLICATION_TYPE',
 'AFFILIATION',
 'CLASSIFICATION',
 'USE_CASE',
 'ORGANIZATION',
 'INCOME_AMT',
 'SPECIAL_CONSIDERATIONS']

In [7]:
# Create a OneHotEncoder instance
enc = OneHotEncoder(sparse = False)

In [8]:
# Encode the categorcal variables using OneHotEncoder
encoded_data = enc.fit_transform(applicant_data_df[categorical_variables])

C:\Users\ppate\anaconda3\Lib\site-packages\sklearn\preprocessing\_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(
C:\Users\ppate\anaconda3\Lib\site-packages\sklearn\utils\validation.py:605: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if is_sparse(pd_dtype):
C:\Users\ppate\anaconda3\Lib\site-packages\sklearn\utils\validation.py:614: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if is_sparse(pd_dtype) or not is_extension_array_dtype(pd_dtype):
C:\Users\ppate\anaconda3\Lib\site-packages\sklearn\utils\validation.py:605: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if is_spar

In [9]:
# Create a DataFrame with the encoded variables
encoded_df = pd.DataFrame(encoded_data,
                          columns = enc.get_feature_names_out(
                              categorical_variables))
encoded_df

APPLICATION_TYPE_T10  APPLICATION_TYPE_T12  APPLICATION_TYPE_T13  \
0                       1.0                   0.0                   0.0   
1                       0.0                   0.0                   0.0   
2                       0.0                   0.0                   0.0   
3                       0.0                   0.0                   0.0   
4                       0.0                   0.0                   0.0   
...                     ...                   ...                   ...   
34294                   0.0                   0.0                   0.0   
34295                   0.0                   0.0                   0.0   
34296                   0.0                   0.0                   0.0   
34297                   0.0                   0.0                   0.0   
34298                   0.0                   0.0                   0.0   

       APPLICATION_TYPE_T14  APPLICATION_TYPE_T15  APPLICATION_TYPE_T17  \
0                       0.0                   0.0                   0.0   
1                       0.0                   0.0                   0.0   
2                       0.0                   0.0                   0.0   
3                       0.0                   0.0                   0.0   
4                       0.0                   0.0                   0.0   
...                     ...                   ...                   ...   
34294                   0.0                   0.0                   0.0   
34295                   0.0                   0.0                   0.0   
34296                   0.0                   0.0                   0.0   
34297                   0.0                   0.0                   0.0   
34298                   0.0                   0.0                   0.0   

       APPLICATION_TYPE_T19  APPLICATION_TYPE_T2  APPLICATION_TYPE_T25  \
0                       0.0                  0.0                   0.0   
1                       0.0                  0.0                   0.0   
2                       0.0                  0.0                   0.0   
3                       0.0                  0.0                   0.0   
4                       0.0                  0.0                   0.0   
...                     ...                  ...                   ...   
34294                   0.0                  0.0                   0.0   
34295                   0.0                  0.0                   0.0   
34296                   0.0                  0.0                   0.0   
34297                   0.0                  0.0                   0.0   
34298                   0.0                  0.0                   0.0   

       APPLICATION_TYPE_T29  ...  INCOME_AMT_1-9999  INCOME_AMT_10000-24999  \
0                       0.0  ...                0.0                     0.0   
1                       0.0  ...                1.0                     0.0   
2                       0.0  ...                0.0                     0.0   
3                       0.0  ...                0.0                     1.0   
4                       0.0  ...                0.0                     0.0   
...                     ...  ...                ...                     ...   
34294                   0.0  ...                0.0                     0.0   
34295                   0.0  ...                0.0                     0.0   
34296                   0.0  ...                0.0                     0.0   
34297                   0.0  ...                0.0                     0.0   
34298                   0.0  ...                0.0                     0.0   

       INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  \
0                           0.0                 0.0               0.0   
1                           0.0                 0.0               0.0   
2                           0.0                 0.0               0.0   
3                           0.0                 0.0               0.0   
4 

### Step 4: Add the original DataFrame’s numerical variables to the DataFrame containing the encoded variables.

> **Note** To complete this step, you will employ the Pandas `concat()` function that was introduced earlier in this course. 

In [10]:
# Add the numerical variables from the original DataFrame to the one-hot 
# encoding DataFrame
encoded_df = pd.concat([applicant_data_df[['STATUS',
                                           'ASK_AMT',
                                           'IS_SUCCESSFUL']],
                        encoded_df], 
                       axis = 1)
encoded_df

STATUS   ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_T10  \
0           1      5000              1                   1.0   
1           1    108590              1                   0.0   
2           1      5000              0                   0.0   
3           1      6692              1                   0.0   
4           1    142590              1                   0.0   
...       ...       ...            ...                   ...   
34294       1      5000              0                   0.0   
34295       1      5000              0                   0.0   
34296       1      5000              0                   0.0   
34297       1      5000              1                   0.0   
34298       1  36500179              0                   0.0   

       APPLICATION_TYPE_T12  APPLICATION_TYPE_T13  APPLICATION_TYPE_T14  \
0                       0.0                   0.0                   0.0   
1                       0.0                   0.0                   0.0   
2                       0.0                   0.0                   0.0   
3                       0.0                   0.0                   0.0   
4                       0.0                   0.0                   0.0   
...                     ...                   ...                   ...   
34294                   0.0                   0.0                   0.0   
34295                   0.0                   0.0                   0.0   
34296                   0.0                   0.0                   0.0   
34297                   0.0                   0.0                   0.0   
34298                   0.0                   0.0                   0.0   

       APPLICATION_TYPE_T15  APPLICATION_TYPE_T17  APPLICATION_TYPE_T19  ...  \
0                       0.0                   0.0                   0.0  ...   
1                       0.0                   0.0                   0.0  ...   
2                       0.0                   0.0                   0.0  ...   
3                       0.0                   0.0                   0.0  ...   
4                       0.0                   0.0                   0.0  ...   
...                     ...                   ...                   ...  ...   
34294                   0.0                   0.0                   0.0  ...   
34295                   0.0                   0.0                   0.0  ...   
34296                   0.0                   0.0                   0.0  ...   
34297                   0.0                   0.0                   0.0  ...   
34298                   0.0                   0.0                   0.0  ...   

       INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                    0.0                     0.0                       0.0   
1                    1.0                     0.0                       0.0   
2                    0.0                     0.0                       0.0   
3                    0.0                     1.0                       0.0   
4                    0.0                     0.0                       1.0   
...                  ...                     ...                       ...   
34294                0.0                     0.0                       0.0   
34295                0.0                     0.0                       0.0   
34296                0.0                     0.0                       0.0   
34297                0.0                     0.0                       0.0   
34298                0.0                     0.0                       0.0   

       INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                     0.0               0.0                     0.0   
1                     0.0               0.0                     0.0   
2                     0.0               0.0                     0.0   
3                     0.0               0.0                     0.0   
4                     0.0               0.0                     0.0   
...          

### Step 5: Using the preprocessed data, create the features (`X`) and target (`y`) datasets. 

In [11]:
# Define the target set y using the IS_SUCCESSFUL column, and review "y"
y = encoded_df["IS_SUCCESSFUL"]
y

0        1
1        1
2        0
3        1
4        1
        ..
34294    0
34295    0
34296    0
34297    1
34298    0
Name: IS_SUCCESSFUL, Length: 34299, dtype: int64

In [12]:
# Define features set X by selecting all columns but IS_SUCCESSFUL, and review
# "X"
X = encoded_df.drop(columns = ["IS_SUCCESSFUL"])
X

STATUS   ASK_AMT  APPLICATION_TYPE_T10  APPLICATION_TYPE_T12  \
0           1      5000                   1.0                   0.0   
1           1    108590                   0.0                   0.0   
2           1      5000                   0.0                   0.0   
3           1      6692                   0.0                   0.0   
4           1    142590                   0.0                   0.0   
...       ...       ...                   ...                   ...   
34294       1      5000                   0.0                   0.0   
34295       1      5000                   0.0                   0.0   
34296       1      5000                   0.0                   0.0   
34297       1      5000                   0.0                   0.0   
34298       1  36500179                   0.0                   0.0   

       APPLICATION_TYPE_T13  APPLICATION_TYPE_T14  APPLICATION_TYPE_T15  \
0                       0.0                   0.0                   0.0   
1                       0.0                   0.0                   0.0   
2                       0.0                   0.0                   0.0   
3                       0.0                   0.0                   0.0   
4                       0.0                   0.0                   0.0   
...                     ...                   ...                   ...   
34294                   0.0                   0.0                   0.0   
34295                   0.0                   0.0                   0.0   
34296                   0.0                   0.0                   0.0   
34297                   0.0                   0.0                   0.0   
34298                   0.0                   0.0                   0.0   

       APPLICATION_TYPE_T17  APPLICATION_TYPE_T19  APPLICATION_TYPE_T2  ...  \
0                       0.0                   0.0                  0.0  ...   
1                       0.0                   0.0                  0.0  ...   
2                       0.0                   0.0                  0.0  ...   
3                       0.0                   0.0                  0.0  ...   
4                       0.0                   0.0                  0.0  ...   
...                     ...                   ...                  ...  ...   
34294                   0.0                   0.0                  0.0  ...   
34295                   0.0                   0.0                  0.0  ...   
34296                   0.0                   0.0                  0.0  ...   
34297                   0.0                   0.0                  0.0  ...   
34298                   0.0                   0.0                  0.0  ...   

       INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                    0.0                     0.0                       0.0   
1                    1.0                     0.0                       0.0   
2                    0.0                     0.0                       0.0   
3                    0.0                     1.0                       0.0   
4                    0.0                     0.0                       1.0   
...                  ...                     ...                       ...   
34294                0.0                     0.0                       0.0   
34295                0.0                     0.0                       0.0   
34296                0.0                     0.0                       0.0   
34297                0.0                     0.0                       0.0   
34298                0.0                     0.0                       0.0   

       INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                     0.0               0.0                     0.0   
1                     0.0               0.0                     0.0   
2                     0.0               0.0                     0.0   
3                     0.0               0.0                     0.0   
4           

### Step 6: Split the features and target sets into training and testing datasets.


In [13]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 1)

### Step 7: Use scikit-learn's `StandardScaler` to scale the features data.

In [14]:
# Create a StandardScaler instance
scaler = StandardScaler()

# Fit the scaler to the features training dataset
X_scaler = scaler.fit(X_train)

X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

C:\Users\ppate\anaconda3\Lib\site-packages\sklearn\utils\validation.py:767: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if not hasattr(array, "sparse") and array.dtypes.apply(is_sparse).any():
C:\Users\ppate\anaconda3\Lib\site-packages\sklearn\utils\validation.py:605: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if is_sparse(pd_dtype):
C:\Users\ppate\anaconda3\Lib\site-packages\sklearn\utils\validation.py:614: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if is_sparse(pd_dtype) or not is_extension_array_dtype(pd_dtype):
C:\Users\ppate\anaconda3\Lib\site-packages\sklearn\utils\validation.py:767: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  i

---

## Compile and Evaluate a Binary Classification Model Using a Neural Network

### Step 1: Create a deep neural network by assigning the number of input features, the number of layers, and the number of neurons on each layer using Tensorflow’s Keras.

> **Hint** You can start with a two-layer deep neural network model that uses the `relu` activation function for both layers.


In [15]:
# Define the the number of inputs (features) to the model
number_input_features = len(X_train.iloc[0]) 
number_input_features

116

In [16]:
# Define the number of neurons in the output layer
number_output_neurons = 1

In [17]:
# Define the number of hidden nodes for the first hidden layer
hidden_nodes_layer1 = (number_input_features + number_output_neurons) // 2
hidden_nodes_layer1

58

In [18]:
# Define the number of hidden nodes for the second hidden layer
hidden_nodes_layer2 = (hidden_nodes_layer1 + number_output_neurons) // 2
hidden_nodes_layer2

29

In [19]:
# Create the Sequential model instance
neural_network = Sequential()

In [20]:
# Add the first hidden layer
neural_network.add(Dense(
    units = hidden_nodes_layer1,
    activation = "relu",
    input_dim = number_input_features))

In [21]:
# Add the second hidden layer
neural_network.add(Dense(
    units = hidden_nodes_layer2,
    activation = "relu",
    input_dim = number_input_features))

In [22]:
# Add the output layer to the model 
neural_network.add(Dense(units = number_output_neurons, activation = "sigmoid"))

In [23]:
# Display the Sequential model summary
neural_network.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 58)                6786      
                                                                 
 dense_1 (Dense)             (None, 29)                1711      
                                                                 
 dense_2 (Dense)             (None, 1)                 30        
                                                                 
Total params: 8527 (33.31 KB)
Trainable params: 8527 (33.31 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


### Step 2: Compile and fit the model using the `binary_crossentropy` loss function, the `adam` optimizer, and the `accuracy` evaluation metric.


In [24]:
# Compile the Sequential model
neural_network.compile(loss="binary_crossentropy", 
                             optimizer="adam",
                             metrics=["accuracy"])

In [25]:
# Fit the model using 50 epochs and the training data
model = neural_network.fit(X_train_scaled, y_train, epochs = 50)

Epoch 1/50


804/804 [==============================] - 5s 3ms/step - loss: 0.5757 - accuracy: 0.7187
Epoch 2/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5530 - accuracy: 0.7292
Epoch 3/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5490 - accuracy: 0.7319
Epoch 4/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5471 - accuracy: 0.7320
Epoch 5/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5457 - accuracy: 0.7338
Epoch 6/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5445 - accuracy: 0.7334
Epoch 7/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5438 - accuracy: 0.7338
Epoch 8/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5433 - accuracy: 0.7344
Epoch 9/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5419 - accuracy: 0.7356
Epoch 10/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5413 - accuracy: 0.73

### Step 3: Evaluate the model using the test data to determine the model’s loss and accuracy.


In [26]:
# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
model_loss, model_accuracy = neural_network.evaluate(X_test_scaled, y_test, verbose = 2)

# Display the model loss and accuracy results
print(f"Loss: {model_loss:.3f}, Accuracy: {model_accuracy:.2f}%")

268/268 - 1s - loss: 0.5564 - accuracy: 0.7282 - 1s/epoch - 4ms/step
Loss: 0.556, Accuracy: 0.73%


---
### The accuracy of the initial model was evaluated at 73%
---

In [77]:
# Save and export your model to an HDF5 file
neural_network.save(Path("initial_model.h5"))

---

## Optimize the neural network model


### Alternative Model 1 - For the first optimization attempt I increased the number of epochs to 100, which was previously 50. 

In [38]:
# Define the the number of inputs (features) to the model
number_input_features_A1 = len(X_train.iloc[0])
number_input_features_A1

116

In [39]:
# Define the number of neurons in the output layer
number_output_neurons_A1 = 1

In [43]:
# Define the number of hidden nodes for the first hidden layer
hidden_nodes_layer1_A1 = (number_input_features_A1 + number_output_neurons_A1) // 2
hidden_nodes_layer1_A1

58

In [44]:
# Define the number of hidden nodes for the second hidden layer
hidden_nodes_layer2_A1 = (hidden_nodes_layer1_A1 + number_output_neurons_A1) // 2
hidden_nodes_layer2_A1

29

In [45]:
# Create the Sequential model instance
neural_network_A1 = Sequential()

In [46]:
# First hidden layer
neural_network_A1.add(Dense(
    units = hidden_nodes_layer1_A1,
    activation = "relu",
    input_dim = number_input_features))

# Add the second hidden layer
neural_network_A1.add(Dense(
    units = hidden_nodes_layer2_A1,
    activation = "relu",
    input_dim = number_input_features))

# Output layer
neural_network_A1.add(Dense(units = number_output_neurons, activation = "sigmoid"))

In [47]:
# Display the Sequential model summary
neural_network_A1.summary()

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_7 (Dense)             (None, 58)                6786      
                                                                 
 dense_8 (Dense)             (None, 29)                1711      
                                                                 
 dense_9 (Dense)             (None, 1)                 30        
                                                                 
Total params: 8527 (33.31 KB)
Trainable params: 8527 (33.31 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [48]:
# Compile the Sequential model
neural_network_A1.compile(loss="binary_crossentropy", 
                             optimizer="adam",
                             metrics=["accuracy"])

In [49]:
# Fit the model using 100 epochs and the training data
fit_model_A1 = neural_network_A1.fit(X_train_scaled, y_train, epochs = 200)

Epoch 1/200
804/804 [==============================] - 4s 3ms/step - loss: 0.5753 - accuracy: 0.7189
Epoch 2/200
804/804 [==============================] - 2s 2ms/step - loss: 0.5539 - accuracy: 0.7288
Epoch 3/200
804/804 [==============================] - 2s 2ms/step - loss: 0.5490 - accuracy: 0.7311
Epoch 4/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5477 - accuracy: 0.7322
Epoch 5/200
804/804 [==============================] - 3s 3ms/step - loss: 0.5465 - accuracy: 0.7318
Epoch 6/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5460 - accuracy: 0.7339
Epoch 7/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5438 - accuracy: 0.7341
Epoch 8/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5440 - accuracy: 0.7328
Epoch 9/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5431 - accuracy: 0.7348
Epoch 10/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5421 - accura

In [50]:
# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
model_loss_A1, model_accuracy_A1 = neural_network_A1.evaluate(X_test_scaled, y_test, verbose = 2)

# Display the model loss and accuracy results
print(f"Loss: {model_loss_A1:.2f}, Accuracy: {model_accuracy_A1:.2f}%")

268/268 - 1s - loss: 0.5670 - accuracy: 0.7285 - 759ms/epoch - 3ms/step
Loss: 0.57, Accuracy: 0.73%


---
### Increasing the number of epochs had a very minimal impact on the accuracy of the model. Accuracy improved from 72.82% to 72.85%.
---

## Alternative Model 2 - An additional hidden layer was added to the model, which previously had two hidden layers. 

In [54]:
# Define the the number of inputs (features) to the model
number_input_features_A2 = len(X_train.iloc[0])
number_input_features_A2

116

In [55]:
# Define the number of neurons in the output layer
number_output_neurons_A2 = 1

In [56]:
# Define the number of hidden nodes for the first hidden layer
hidden_nodes_layer1_A2 = (number_input_features_A2 + number_output_neurons_A2) // 2
hidden_nodes_layer1_A2

58

In [59]:
# Define the number of hidden nodes for the second hidden layer
hidden_nodes_layer2_A2 = (hidden_nodes_layer1_A2 + number_output_neurons_A2) // 2
hidden_nodes_layer2_A2

29

In [60]:
# Define the number of hidden nodes for the third hidden layer
hidden_nodes_layer2_A2 = (hidden_nodes_layer2_A2 + number_output_neurons_A2) // 2
hidden_nodes_layer2_A2

15

In [63]:
# Create the Sequential model instance
neural_network_A2 = Sequential()

In [64]:
# First hidden layer
neural_network_A2.add(Dense(
    units = hidden_nodes_layer1_A2,
    activation = "relu",
    input_dim = number_input_features))

# Add the second hidden layer
neural_network_A2.add(Dense(
    units = hidden_nodes_layer2_A2,
    activation = "relu",
    input_dim = number_input_features))

# Add the third hidden layer
neural_network_A2.add(Dense(
    units = hidden_nodes_layer2_A3,
    activation = "relu",
    input_dim = number_input_features))

# Output layer
neural_network_A2.add(Dense(units = number_output_neurons, activation = "sigmoid"))

In [65]:
# Compile the Sequential model
neural_network_A2.compile(loss="binary_crossentropy", 
                             optimizer="adam",
                             metrics=["accuracy"])

In [70]:
# Fit the model using original 50 epochs and the training data
fit_model_A2 = neural_network_A2.fit(X_train_scaled, y_train, epochs = 50)

Epoch 1/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5251 - accuracy: 0.7419
Epoch 2/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5250 - accuracy: 0.7427
Epoch 3/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5254 - accuracy: 0.7418
Epoch 4/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5250 - accuracy: 0.7418
Epoch 5/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5258 - accuracy: 0.7405
Epoch 6/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5251 - accuracy: 0.7413
Epoch 7/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5250 - accuracy: 0.7422
Epoch 8/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5266 - accuracy: 0.7422
Epoch 9/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5279 - accuracy: 0.7422
Epoch 10/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5248 - accuracy: 0.7425

In [69]:
# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
model_loss_A2, model_accuracy_A2 = neural_network_A2.evaluate(X_test_scaled, y_test, verbose = 2)

# Display the model loss and accuracy results
print(f"Loss: {model_loss_A2:.2f}, Accuracy: {model_accuracy_A2:.2f}%")

268/268 - 1s - loss: 0.5730 - accuracy: 0.7284 - 504ms/epoch - 2ms/step
Loss: 0.57, Accuracy: 0.73%


---
### Using the initial amount of 50 epochs, and adding a hidden layer to the model the accuracy also remained relatively unchanged at 72.98%. The additional layer had a greater impact on the models accuracy than a larger number of epochs.
---

## Display the accuracy scores of each model

In [74]:
print("Original Model Results")
model_loss, model_accuracy = neural_network.evaluate(X_test_scaled, y_test, verbose = 2)
print(f"Loss: {model_loss:.3f}, Accuracy: {model_accuracy:.2f}%")
print("Alternative Model 1 Results")
model_loss_A1, model_accuracy_A1 = neural_network_A1.evaluate(X_test_scaled, y_test, verbose = 2)
print(f"Loss: {model_loss_A1:.2f}, Accuracy: {model_accuracy_A1:.2f}%")
print("Alternative Model 2 Results")
model_loss_A2, model_accuracy_A2 = neural_network_A2.evaluate(X_test_scaled, y_test, verbose = 2)
print(f"Loss: {model_loss_A2:.2f}, Accuracy: {model_accuracy_A2:.2f}%")

Original Model Results
268/268 - 1s - loss: 0.5564 - accuracy: 0.7282 - 642ms/epoch - 2ms/step
Loss: 0.556, Accuracy: 0.73%
Alternative Model 1 Results
268/268 - 1s - loss: 0.5670 - accuracy: 0.7285 - 583ms/epoch - 2ms/step
Loss: 0.57, Accuracy: 0.73%
Alternative Model 2 Results
268/268 - 1s - loss: 0.5813 - accuracy: 0.7287 - 607ms/epoch - 2ms/step
Loss: 0.58, Accuracy: 0.73%


## Save alternative models as HDF5 files


In [76]:
# Save and export your model to an HDF5 file
neural_network_A1.save(Path("alternative_model_A1.h5"))

In [75]:
# Save and export your model to an HDF5 file
neural_network_A2.save(Path("alternative_model_A2.h5"))